File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean findTarget(TreeNode root, int k) {
     Stack<TreeNode> inOrderStack = new Stack<>();
        Stack<TreeNode> revInOrderStack = new Stack<>();
        
        TreeNode leftNode = root;
        TreeNode rightNode = root;

        while (true) {
            // Traverse the left subtree for the smallest values
            while (leftNode != null) {
                inOrderStack.push(leftNode);
                leftNode = leftNode.left;
            }

            // Traverse the right subtree for the largest values
            while (rightNode != null) {
                revInOrderStack.push(rightNode);
                rightNode = rightNode.right;
            }

            if (inOrderStack.isEmpty() || revInOrderStack.isEmpty()) {
                return false; // No valid pairs
            }

            TreeNode minNode = inOrderStack.peek();  // Node with the minimum value
            TreeNode maxNode = revInOrderStack.peek(); // Node with the maximum value

            if (minNode == maxNode) {
                return false; // Same node can't be part of a valid pair
            }

            int sum = minNode.val + maxNode.val;
            if (sum == k) {
                return true; // Found the pair
            } else if (sum < k) {
                // Move to the next larger node in the in-order traversal
                inOrderStack.pop();
                leftNode = minNode.right;
            } else {
                // Move to the next smaller node in the reverse in-order traversal
                revInOrderStack.pop();
                rightNode = maxNode.left;
            }
        }
    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        
        stack<TreeNode*> inord, revInord;
        TreeNode* root1 = root;
        TreeNode* root2 = root;
        
        while(true){
            while(root1)    
                inord.push(root1), root1 = root1->left;
            
            while(root2)    
                revInord.push(root2), root2 = root2->right;
            
            // If any of the stacks is empty, break the loop as there's no further exploration possible
            if(inord.empty() || revInord.empty())   
                break;
            
            root1 = inord.top();    // Node with the minimum value
            root2 = revInord.top(); // Node with the maximum value.
            
            if(root1 == root2) {
                // If the same node is found in both stacks, break the loop as no further exploration is needed
                break;
            }
            
            if(root1->val + root2->val == k){
                // If sum equals k, return true
                return true;
            }
            
            if(root1->val + root2->val < k){
                // If sum is less than k, pop the node with minimum value and move to its right subtree
                inord.pop();
                root1 = root1->right;
                root2 = nullptr; // Reset root2 to nullptr for next iteration
            }
            else{
                // If sum is greater than k, pop the node with maximum value and move to its left subtree
                revInord.pop();
                root2 = root2->left;
                root1 = nullptr; // Reset root1 to nullptr for next iteration
            }
        }
        // If loop breaks, it means no such pair exists
        return false;
    }
};
